### ISSUE TITLE
mod_vcard_ldap crashes in 14.12 and 15.02

### ISSUE THREAD

**Issue by Estartu** [processone/ejabberd#494]
2015-03-25 13:46:18 =CRASH REPORT====
  crasher:
    initial call: eldap:init/1
    pid: <0.892.0>
    registered_name: 'eldap_#Ref<0.0.0.19880>'
    exception exit:
{{{badmatch,{error,{asn1,{function_clause,[{'ELDAPv3',enc_SubstringFilter_substrings_components,[{'SubstringFilter_substrings',[{any,<<"estartu">>}]},[],0],[{file,"src/ELDAPv3.erl"},{line,1768}]},{'ELDAPv3',enc_SubstringFilter_substrings,2,[{file,"src/ELDAPv3.erl"},{line,1765}]},{'ELDAPv3',enc_SubstringFilter,2,[{file,"src/ELDAPv3.erl"},{line,1753}]},{'ELDAPv3',enc_Filter,2,[{file,"src/ELDAPv3.erl"},{line,1604}]},{'ELDAPv3',enc_Filter_and_components,3,[{file,"src/ELDAPv3.erl"},{line,1636}]},{'ELDAPv3',enc_Filter_and,2,[{file,"src/ELDAPv3.erl"},{line,1629}]},{'ELDAPv3',enc_Filter,2,[{file,"src/ELDAPv3.erl"},{line,1596}]},{'ELDAPv3',enc_SearchRequest,2,[{file,"src/ELDAPv3.erl"},{line,1511}]}]}}}},[{eldap,send_command,3,[{file,"src/eldap.erl"},{line,831}]},{eldap,process_command,3,[{file,"src/eldap.erl"},{line,813}]},{gen_fsm,handle_msg,7,[{file,"gen_fsm.erl"},{line,503}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]},[{gen_fsm,terminate,7,[{file,"gen_fsm.erl"},{line,611}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]}
    ancestors: ['ejabberd_mod_vcard_ldap_ze.tum.de',ejabberd_sup,<0.37.0>]
    messages: []
    links: [<0.889.0>,#Port<0.6243>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 1290
  neighbours:
    neighbour:
[{pid,<0.889.0>},{registered_name,'ejabberd_mod_vcard_ldap_ze.tum.de'},{initial_call,{mod_vcard_ldap,init,['Argument__1']}},
{current_function,{gen_server,loop,6}},{ancestors,[ejabberd_sup,<0.37.0>]},{messages,[]},{links,[<0.299.0>,<0.892.0>]},{dictionary,[]},{trap_exit,false},{status,waiting},{heap_size,1598},{stack_size,9},{reductions,1139}]
2015-03-25 13:46:18 =SUPERVISOR REPORT====
     Supervisor: {local,ejabberd_sup}
     Context:    child_terminated
     Reason:
{{badmatch,{error,{asn1,{function_clause,[{'ELDAPv3',enc_SubstringFilter_substrings_components,[{'SubstringFilter_substrings',[{any,<<"estartu">>}]},[],0],[{file,"src/ELDAPv3.erl"},{line,1768}]},{'ELDAPv3',enc_SubstringFilter_substrings,2,[{file,"src/ELDAPv3.erl"},{line,1765}]},{'ELDAPv3',enc_SubstringFilter,2,[{file,"src/ELDAPv3.erl"},{line,1753}]},{'ELDAPv3',enc_Filter,2,[{file,"src/ELDAPv3.erl"},{line,1604}]},{'ELDAPv3',enc_Filter_and_components,3,[{file,"src/ELDAPv3.erl"},{line,1636}]},{'ELDAPv3',enc_Filter_and,2,[{file,"src/ELDAPv3.erl"},{line,1629}]},{'ELDAPv3',enc_Filter,2,[{file,"src/ELDAPv3.erl"},{line,1596}]},{'ELDAPv3',enc_SearchRequest,2,[{file,"src/ELDAPv3.erl"},{line,1511}]}]}}}},[{eldap,send_command,3,[{file,"src/eldap.erl"},{line,831}]},{eldap,process_command,3,[{file,"src/eldap.erl"},{line,813}]},{gen_fsm,handle_msg,7,[{file,"gen_fsm.erl"},{line,503}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]}
     Offender:
[{pid,<0.889.0>},{name,'ejabberd_mod_vcard_ldap_ze.tum.de'},{mfargs,{mod_vcard_ldap,start_link,[<<"ze.tum.de">>,[{search,true},{ldap_vcard_map,[{<<"NICKNAME">>,[{<<"%u">>,[]}]},{<<"FN">>,[{<<"%s">>,[<<"gecos">>]}]},{<<"EMAIL">>,[{<<"%u@%d">>,[]}]}]},{ldap_searc
h_fields,[{<<"FullName">>,<<"gecos">>},{<<"Nickname">>,<<"%u">>}]},{ldap_search_reported,[{<<"FullName">>,<<"FN">>},{<<"Nickname">>,<<"NICKNAME">>}]}]]}},{restart_type,transient},{shutdown,1000},{child_type,worker}]

The config is
    ldap_vcard_map:
      "NICKNAME":
        "%u": []
      "FN":
        "%s":
          - "gecos"
      "EMAIL":
        "%u@%d": []
    ldap_search_fields:
      "Full Name": "gecos"
      "Nickname": "%u"
    ldap_search_reported:
      "Full Name": "FN"
      "Nickname": "NICKNAME"

Authentication agains ldap works just fine.

Regards
   Estartu

**Comment by brunoqc** [comment]
I also have this problem with 15.03.

**Comment by yankaiqian** [comment]
I had a similar issue, posted in the forum, https://www.ejabberd.im/node/24837

I get it work by changing the ldap_uids setting in the authentication section, not sure if it helps, just for your reference.

**Comment by brunoqc** [comment]
I have the crash with the following setting. I'm able to get my vcard but I get the crash when I search for a user.

```yaml
ldap_uids:
  "mail": "%u@domain.com"
```

```
2015-04-07 10:40:34 =ERROR REPORT====
** State machine 'eldap_#Ref<0.0.0.10799>' terminating
** Last message in was {'$gen_sync_event',{<0.482.0>,#Ref<0.0.0.10834>},{search,{eldap_search,wholeSubtree,<<"ou=LUXO,o=NET">>,{and,[{and,[{substrings,{'SubstringFilter',<<"mail">>,{'SubstringFilter_substrings',[{final,<<"@domain.com">>}]}}},{equalityMatch,{'AttributeValueAssertion',<<"objectClass">>,<<"Person">>}}]},{and,[]}]},30,[<<"birthDay">>,<<"c">>,<<"displayName">>,<<"initials">>,<<"l">>,<<"mail">>,<<"o">>,<<"ou">>],false,neverDerefAliases,0}}}
** When State == active
**      Data  == {eldap,3,[<<"ldap.domain.com">>],<<"ldap.domain.com">>,636,ssl,tls,[],{sslsocket,{gen_tcp,#Port<0.3956>,tls_connection},<0.480.0>},<<>>,<<>>,1,#Ref<0.0.0.10824>,{dict,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{[],[]}}
** Reason for termination =
** {{badmatch,{error,{asn1,{function_clause,[{'ELDAPv3',enc_SubstringFilter_substrings_components,[{'SubstringFilter_substrings',[{final,<<"@domain.com">>}]},[],0],[{file,"src/ELDAPv3.erl"},{line,1768}]},{'ELDAPv3',enc_SubstringFilter_substrings,2,[{file,"src/ELDAPv3.erl"},{line,1765}]},{'ELDAPv3',enc_SubstringFilter,2,[{file,"src/ELDAPv3.erl"},{line,1753}]},{'ELDAPv3',enc_Filter,2,[{file,"src/ELDAPv3.erl"},{line,1604}]},{'ELDAPv3',enc_Filter_and_components,3,[{file,"src/ELDAPv3.erl"},{line,1636}]},{'ELDAPv3',enc_Filter_and,2,[{file,"src/ELDAPv3.erl"},{line,1629}]},{'ELDAPv3',enc_Filter,2,[{file,"src/ELDAPv3.erl"},{line,1596}]},{'ELDAPv3',enc_Filter_and_components,3,[{file,"src/ELDAPv3.erl"},{line,1636}]}]}}}},[{eldap,send_command,3,[{file,"src/eldap.erl"},{line,831}]},{eldap,process_command,3,[{file,"src/eldap.erl"},{line,813}]},{gen_fsm,handle_msg,7,[{file,"gen_fsm.erl"},{line,505}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,239}]}]}
2015-04-07 10:40:34 =CRASH REPORT====
  crasher:
    initial call: eldap:init/1
    pid: <0.479.0>
    registered_name: 'eldap_#Ref<0.0.0.10799>'
    exception exit: {{{badmatch,{error,{asn1,{function_clause,[{'ELDAPv3',enc_SubstringFilter_substrings_components,[{'SubstringFilter_substrings',[{final,<<"@domain.com">>}]},[],0],[{file,"src/ELDAPv3.erl"},{line,1768}]},{'ELDAPv3',enc_SubstringFilter_substrings,2,[{file,"src/ELDAPv3.erl"},{line,1765}]},{'ELDAPv3',enc_SubstringFilter,2,[{file,"src/ELDAPv3.erl"},{line,1753}]},{'ELDAPv3',enc_Filter,2,[{file,"src/ELDAPv3.erl"},{line,1604}]},{'ELDAPv3',enc_Filter_and_components,3,[{file,"src/ELDAPv3.erl"},{line,1636}]},{'ELDAPv3',enc_Filter_and,2,[{file,"src/ELDAPv3.erl"},{line,1629}]},{'ELDAPv3',enc_Filter,2,[{file,"src/ELDAPv3.erl"},{line,1596}]},{'ELDAPv3',enc_Filter_and_components,3,[{file,"src/ELDAPv3.erl"},{line,1636}]}]}}}},[{eldap,send_command,3,[{file,"src/eldap.erl"},{line,831}]},{eldap,process_command,3,[{file,"src/eldap.erl"},{line,813}]},{gen_fsm,handle_msg,7,[{file,"gen_fsm.erl"},{line,505}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,239}]}]},[{gen_fsm,terminate,7,[{file,"gen_fsm.erl"},{line,622}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,239}]}]}
    ancestors: ['ejabberd_mod_vcard_ldap_xmpp.app.local',ejabberd_sup,<0.38.0>]
    messages: []
    links: [<0.476.0>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 1984
  neighbours:
    neighbour: [{pid,<0.476.0>},{registered_name,'ejabberd_mod_vcard_ldap_xmpp.app.local'},{initial_call,{mod_vcard_ldap,init,['Argument__1']}},{current_function,{gen_server,loop,6}},{ancestors,[ejabberd_sup,<0.38.0>]},{messages,[]},{links,[<0.310.0>,<0.479.0>]},{dictionary,[]},{trap_exit,false},{status,waiting},{heap_size,1598},{stack_size,9},{reductions,1162}]
2015-04-07 10:40:34 =SUPERVISOR REPORT====
     Supervisor: {local,ejabberd_sup}
     Context:    child_terminated
     Reason:     {{badmatch,{error,{asn1,{function_clause,[{'ELDAPv3',enc_SubstringFilter_substrings_components,[{'SubstringFilter_substrings',[{final,<<"@domain.com">>}]},[],0],[{file,"src/ELDAPv3.erl"},{line,1768}]},{'ELDAPv3',enc_SubstringFilter_substrings,2,[{file,"src/ELDAPv3.erl"},{line,1765}]},{'ELDAPv3',enc_SubstringFilter,2,[{file,"src/ELDAPv3.erl"},{line,1753}]},{'ELDAPv3',enc_Filter,2,[{file,"src/ELDAPv3.erl"},{line,1604}]},{'ELDAPv3',enc_Filter_and_components,3,[{file,"src/ELDAPv3.erl"},{line,1636}]},{'ELDAPv3',enc_Filter_and,2,[{file,"src/ELDAPv3.erl"},{line,1629}]},{'ELDAPv3',enc_Filter,2,[{file,"src/ELDAPv3.erl"},{line,1596}]},{'ELDAPv3',enc_Filter_and_components,3,[{file,"src/ELDAPv3.erl"},{line,1636}]}]}}}},[{eldap,send_command,3,[{file,"src/eldap.erl"},{line,831}]},{eldap,process_command,3,[{file,"src/eldap.erl"},{line,813}]},{gen_fsm,handle_msg,7,[{file,"gen_fsm.erl"},{line,505}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,239}]}]}
     Offender:   [{pid,<0.476.0>},{name,'ejabberd_mod_vcard_ldap_xmpp.app.local'},{mfargs,{mod_vcard_ldap,start_link,[<<"xmpp.app.local">>,[]]}},{restart_type,transient},{shutdown,1000},{child_type,worker}]
```

**Comment by zinid** [comment]
This was fixed in ed79a61752dc137f83c8bbd8963924baea53ddc9
